engine/common/version/version_control.go

```diff
@@ -0,0 +1,364 @@
+package version
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/counters"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+	psEvents "github.com/onflow/flow-go/state/protocol/events"
+	"github.com/onflow/flow-go/storage"
+)
+
+// ErrOutOfRange indicates that height is higher than last handled block height
+var ErrOutOfRange = errors.New("height is out of range")
+
+// VersionControlConsumer defines a function type that consumes version control updates.
+// It is called with the block height and the corresponding semantic version.
+// There are two possible notifications options:
+// - An updated version will have a new height and a semantic version at that height.
+// - A deleted version will have the previous height and an empty semantic version, indicating that the update was deleted.
+type VersionControlConsumer func(height uint64, version *semver.Version)
```

Comment:
Since we're passing a pointer now, I think `nil` makes the most sense. using `0.0.0` requires consumers do semver comparison instead of just a nil check
```suggestion
// - A new or updated version will have a height and a semantic version at that height.
// - A deleted version will have the previous height and nil semantic version, indicating that the update was deleted.
type VersionControlConsumer func(height uint64, version *semver.Version)
```